### PR TITLE
feat: Add support for removing fields from diff

### PR DESCRIPTION
**Pull request by mdnfiras**
related: https://github.com/nolar/kopf/issues/1087

This PR solves [this comment](https://github.com/nolar/kopf/issues/1087#issuecomment-2186077417) from the previously mentioned issue.

[trivy-dojo-report-operator](https://github.com/telekom-mms/trivy-dojo-report-operator)'s `VulnerabilityReport` resources can get very big and cause issues. Having the ability to remove a couple of fields from the diff solves it.

Acknowledgement: Thanks to @baracoder for suggesting the core idea behind this approach!

### CODE REVIEW

File: kopf/_cogs/configs/diffbase.py [hunk 1]
```diff
@@ -111,10 +111,12 @@ def __init__(
             *,
             prefix: str = 'kopf.zalando.org',
             key: str = 'last-handled-configuration',
+            remove_fields: Iterable[str] = None,
```
Comment (nolar) [inline kopf/_cogs/configs/diffbase.py]: Can you please rename it to a single-word `ignored_fields`?

So that it is aligned with `extra_fields` in the same module & class, following the "adjective + noun" semantics instead of the imperative verb. (Any other appropriate adjective would work, too.)

File: kopf/_cogs/configs/diffbase.py [hunk 2]
```diff
@@ -123,6 +125,18 @@ def build(
             extra_fields: Iterable[dicts.FieldSpec] | None = None,
     ) -> bodies.BodyEssence:
         essence = super().build(body=body, extra_fields=extra_fields)
+
+        # Remove extra fields if specified
+        for path in self._remove_fields:
+            d = essence
+            keys = path.split('.')
```
Comment (nolar) [inline kopf/_cogs/configs/diffbase.py]: ~It is better to use `dicts.parse_field()` from `kopf._cogs.structs.dicts` — to keep the field specification conventions unified across the interfaces.~

Probably not relevant, since I recommended using `dicts.remove()` below, which accepts unparsed field specs.

File: kopf/_cogs/configs/diffbase.py (hunk 1 above)
Comment (nolar) [inline kopf/_cogs/configs/diffbase.py]: It is better to use `dicts.FieldSpec` from `kopf._cogs.structs.dicts` — to keep the field specification conventions unified across the interfaces. This also include the pre-parsed lists and tuples of individual field components.

`dicts` is already imported in this file.

File: kopf/_cogs/configs/diffbase.py [hunk 3]
```diff
@@ -183,6 +199,17 @@ def build(
...
+            for k in keys[:-1]:
+                # If current parent field is not a dict then the path is invalid; skip removal
+                if not isinstance(d.get(k, {}), dict):
+                    break
+                d = d.get(k, {})
+            d.pop(keys[-1], None)
+
```
Comment (nolar) [inline kopf/_cogs/configs/diffbase.py]: Should this functionality be moved to the parent class `DiffBaseStorage`, in order to avoid the code duplication?

File: kopf/_cogs/configs/diffbase.py [hunk 4]
```diff
@@ -123,6 +125,18 @@ def build(
...
+        # Remove extra fields if specified
+        for path in self._remove_fields:
+            d = essence
+            keys = path.split('.')
+            for k in keys[:-1]:
+                # If current parent field is not a dict then the path is invalid; skip removal
+                if not isinstance(d.get(k, {}), dict):
+                    break
+                d = d.get(k, {})
+            d.pop(keys[-1], None)
```
Comment (nolar) [inline kopf/_cogs/configs/diffbase.py]: I have a strong suspicion that this code semantically duplicates `dicts.remove()` from `kopf._cogs.structs.dicts`, which has an advantage of also removing the empty parents.

In that case, you can avoid writing numerous edge-case tests for the fields removal, as the removal is tested. Only a single test that the removal is happening/attempted at all would suffice.

File: tests/persistence/test_storing_of_diffbase.py [hunk 5]
```diff
@@ -65,11 +65,21 @@ def test_annotations_storage_with_prefix_and_key():
     assert storage.prefix == 'my-operator.my-company.com'
     assert storage.key == 'diff-base'
 
+def test_annotations_storage_with_remove_fields():
+    storage = AnnotationsDiffBaseStorage(
+        remove_fields=['a.b', 'a.e.f.g', 'h', 'k.l.m']
+    )
+    assert storage._remove_fields == ['a.b', 'a.e.f.g', 'h', 'k.l.m']
```
Comment (nolar) [inline tests/persistence/test_storing_of_diffbase.py]: It makes no sense on testing the private/protected fields. Depending on the intention, either make it public (as other fields), or do not test it.

However, a test that checks that the removal actually happens or is attempted is needed (as in: "test the promise, not the implementation").

**Review by nolar — CHANGES_REQUESTED** [review]
Thanks for the contribution! The functionality might be helpful indeed. But it is better to use the existing routines to manage the dicts and fields uniformly with the rest of the framework.

File: kopf/_cogs/configs/diffbase.py (hunk 4 above)
Comment (mdnfiras) [reply kopf/_cogs/configs/diffbase.py]: `dicts.remove` produces an error `TypeError: 'str' object does not support item deletion` when the user tries to delete a field that is actually a string (or any type that doesn't support field deletion). For example trying to delete the field `a.b` will produce this error because `a` is a string and has no subfields:
```json
{
  "a": "b"
}
```
In my code I handle it with the check `if not isinstance(d.get(k, {}), dict):` and just skip it. also `d = d.get(k, {})` ensures that non existent fields are also skipped silently.
Do you think I should still catch these errors and handle them silently or let the errors propagate?

ps: thanks for the review!

File: tests/persistence/test_storing_of_diffbase.py (hunk 5 above)
Comment (mdnfiras) [reply tests/persistence/test_storing_of_diffbase.py]: Alright, I don't intend to access this field for now so I'm keeping it protected. I removed this kind of tests for this field.

File: kopf/_cogs/configs/diffbase.py (hunk 4 above)
Comment (mdnfiras) [reply kopf/_cogs/configs/diffbase.py]: I ended up silently ignoring non-existent fields, and fields that don't support item deletion.

Comment (mdnfiras) [reply kopf/_cogs/configs/diffbase.py]: note: `dicts.remove` already ignores non-existent fields, so catching `TypeError` exception is just for ignoring fields that don't support item deletion.

File: tests/persistence/test_essences.py [hunk 6]
```diff
@@ -178,3 +178,33 @@ def test_get_essence_clones_body(
...
+            },
+            'e': {
+                'f': 'g',
+            }
+        },
+        'h': {
+            'i': 'j',
+        }
+    })
+    storage = cls(ignored_fields=['a.b', 'a.e.f.g', 'h', 'k.l.m'])
```
Comment (nolar) [inline tests/persistence/test_essences.py:199]: Just for the record: `a.e.f.g` is not a field, only `a.e.f` is — so it is an extra test for the removal of a non-existent field, the same as `k.l.m`.

File: kopf/_cogs/configs/diffbase.py [hunk 7]
```diff
@@ -26,6 +26,10 @@ class DiffBaseStorage(conventions.StorageKeyMarkingConvention,
     https://kubernetes.io/docs/concepts/overview/object-management-kubectl/declarative-config/
     """
 
+    def __init__(self, ignored_fields: Iterable[dicts.FieldSpec] | None = None) -> None:
+        super().__init__()
+        self.ignored_fields = list(ignored_fields or [])  # materialize the iterable
```
Comment (nolar) [inline kopf/_cogs/configs/diffbase.py:31]: Note: I've added this materialization here, so that depletable iterables would persist and be reusable across multiple runs. I hope nobody uses generators (both expressions & functions) in this context, but just in case.

**Review by nolar — APPROVED** [review]

**Comment by nolar** [comment]
Released as 1.39.1. Thanks for the contribution!

File: tests/persistence/test_essences.py (hunk 6 above)
Comment (mdnfiras) [reply tests/persistence/test_essences.py:199]: Yes `a.e.f.g` is technically not a field, but this one in particular causes `dicts.remove` to throw a `TypeError` exception because `a.e.f` is a field but does not support item deletion, hence the extra test.

Comment (nolar) [reply tests/persistence/test_essences.py:199]: Got it. Makes sense, indeed.
